### ISSUE TITLE
Removed files re-appear when using cache

### ISSUE THREAD

**Issue by orls** [GoogleContainerTools/kaniko#704]
**Actual behavior**

Building a Dockerfile which deletes a file _initially_ produces an image without the file. But when re-built using the cache, the resulting image contains the file again.

**Expected behavior**

The deleted file should never be present in final images, regardless of whether cache was used.

**To Reproduce**

1. Build a simple Dockerfile that adds a file then removes it in a later step, using Kaniko, with caching enabled:

```
FROM alpine

RUN mkdir -p /some/dir/ && echo 'first' > /some/dir/first.txt

RUN rm /some/dir/first.txt
```

2. Pull the resulting image from your registry, and run `ls -la /some/dir` in a container using the image. Observe that the directory is empty:

```
$ docker pull my-registry/my/img:latest && docker run --rm my-registry/my/img:latest bin/sh -c "ls -la /some/dir"
latest: Pulling from my/img
921b31ab772b: Already exists
b4f1d0fd5c4c: Pull complete
f6d9b826d061: Pull complete
Digest: sha256:213ace2d10ba510eeeaefa90b255da5f48e87ff7577a560bbb262b9db1cd9945
Status: Downloaded newer image for my-registry/my/img:latest
my-registry/my/img:latest
total 8
drwxr-xr-x    1 root     root          4096 Jun 24 14:27 .
drwxr-xr-x    1 root     root          4096 Jun 24 14:27 ..
```

3. Re-build the image using Kaniko, exactly as before. Logs will show that cached layers are being used.

4. Re-pull the image, re-run a container to `ls` again. Observe that the file has re-appeared.

```
docker pull my-registry/my/img:latest && docker run --rm my-registry/my/img:latest bin/sh -c "ls -la /some/dir"
latest: Pulling from my/img
921b31ab772b: Already exists
bd7f0d50b1ce: Pull complete
7d792274c478: Pull complete
Digest: sha256:9adda5568544f015972d112a9de87829ee2f02c0f65871ca2c632b10201be096
Status: Downloaded newer image for my-registry/my/img:latest
my-registry/my/img:latest
total 12
drwxr-xr-x    1 root     root          4096 Jun 24 14:34 .
drwxr-xr-x    1 root     root          4096 Jun 24 14:34 ..
-rw-r--r--    1 root     root             6 Jun 24 14:34 first.txt
```

**Additional Information**

- Example Dockerfile above.
- Build context is empty/irrelevant.
- Tried various combinations of `ADD <file-from-context> <path>`, `ADD <archive-of-directory>.tar.gz /`, `COPY <file-from-context> <path>` and using `RUN` steps to manipulate the files (as in the example). All combinations show this behaviour.
- This behaviour has been seen with various other later steps after the deletion (fs-related stuff, non-FS things like `ENTRYPOINT` etc).
- Kaniko Image details:

```
$ docker inspect gcr.io/kaniko-project/executor:v0.10.0
[
    {
        "Id": "sha256:9000801241318d6c9b8c2b505974f22fd74c57e0adfdb09c98870aef4923bb65",
        "RepoTags": [
            "gcr.io/kaniko-project/executor:latest",
            "gcr.io/kaniko-project/executor:v0.10.0"
        ],
        "RepoDigests": [
            "gcr.io/kaniko-project/executor@sha256:78d44ec4e9cb5545d7f85c1924695c89503ded86a59f92c7ae658afa3cff5400"
        ],
        ...
```

I can see from logs that a whiteout file is being created as expected in the first (uncached) build. However, it looks like the whiteout is either not making it's way into the cached data, or isn't being honoured when the cache is used later?

Peeking inside docker's `overlay2` directory confirms the lack of a whiteout in 2nd img. Upper layer internals in 1st image, built *without* cache:

```
/ # ls -la /var/lib/docker/overlay2/596f011a6b72769df25c4bd64e247bcede1f08f43f72be21790fc5a347d43921/diff/some/dir
total 8
drwxr-xr-x    2 root     root          4096 Jun 24 14:27 .
drwxr-xr-x    3 root     root          4096 Jun 24 14:27 ..
c---------    1 root     root        0,   0 Jun 24 14:28 first.txt
```

Upper layer internals for 2nd image, built *with* cache:

```
/ # ls -la /var/lib/docker/overlay2/afd849e49a216c891156d60f0dc99e4a6eb5b2a59ece4ca4d75e3705df9edc9e/diff/some/dir
total 8
drwxr-xr-x    2 root     root          4096 Jun 24 14:34 .
drwxr-xr-x    3 root     root          4096 Jun 24 14:34 ..
```

This behaviour has been observed in multiple environments:

- Using Jenkins on Kubernetes in GCP (as per [official suggestion](https://cloud.google.com/solutions/jenkins-on-kubernetes-engine) & [example](https://github.com/jenkinsci/kubernetes-plugin/blob/kubernetes-1.14.5/examples/kaniko-declarative.groovy)), with gcr.io  as final img dest & a k8s-deployed docker registry for cache.
- On a mac, using Kaniko via docker, internal private registry as final img dest & locally-running registry for cache (for easy testing)

**Comment by orls** [comment]
Digging further, I can see that the whiteout file is present in the cache data (fetched relevant blob tarball  directly from the cache registry):

```
tar -tzvf ./removal-layer-from-cache-registry.tar.gz
----------  0 0      0           0  1 Jan  1970 /some/dir/.wh.first.txt
drwxr-xr-x  0 /root  root        0 24 Jun 16:23 /some
drwxr-xr-x  0 /root  root        0 24 Jun 16:23 /some/dir
drwxr-xr-x  0 /root  root        0 24 Jun 16:23 /
```

**Comment by bymarshall** [comment]
Hi, I confirmed this incorrect behavior. As a result, when kaniko's cache is activated, it increases the size of the containers and probably generates other problems by leaving the files and folders that you want to delete.

**Comment by cvgw** [comment]
This is definitely a bug in how kaniko's caching is implemented. I think the issue most likely stems from the fact that kaniko unpacks layers to the filesystem even when they are cached, but does not execute cached run commands.

```
-> kaniko retrieves cached layer
-> kaniko unpacks cached layer to filesystem
-> kaniko sees RUN rm /some-file is cached and skips the command
-> kaniko packs up the files from the filesystem which includes the file which should have been delete
```

**Comment by cvgw** [comment]
I'm not super familiar with this part of the code but here is what I found

During stage build when the current command is a `CachedRunCommand` we call `s.snapshotter.TakeSnapshot`
https://github.com/GoogleContainerTools/kaniko/blob/732b2700354dc0f501e5d3012e74b3d6672bae41/pkg/executor/build.go#L368

But `Snapshotter.TakeSnapshot` does include the deleted file (whiteout) logic
https://github.com/GoogleContainerTools/kaniko/blob/732b2700354dc0f501e5d3012e74b3d6672bae41/pkg/snapshot/snapshot.go#L63

Because that is only executed by `Snapshotter.TakeSnapshotFS`
https://github.com/GoogleContainerTools/kaniko/blob/732b2700354dc0f501e5d3012e74b3d6672bae41/pkg/snapshot/snapshot.go#L107

the TLDR being that when we have a `RunCommand` we call `Snapshotter.TakeSnapshotFS` which tracks the deleted files. But when we have a `CachedRunCommand` we call `Snapshotter.TakeSnapshot` which doesn't handle the deleted files.

I'm guessing there is a reason why we call one in one case and the other in another rather than `Snapshotter.TakeSnapshotFS` for both cases, but I'm not sure what that is.

**Comment by cvgw** [comment]
You can see that the final layer for the cached and uncached versions of the image are different
```
cgwippern@cgwippern:~/kaniko-dev/issues/704$ tar tvf image-cached/ff195ce13b387856dbe6a447df590a584154dccbc6345e7029dc42879fd0910e/layer.tar --absolute-names
drwxr-xr-x /root/0           0 2019-12-22 13:23 /
drwxr-xr-x /root/0           0 2019-12-22 13:23 some
drwxr-xr-x /root/0           0 2019-12-22 13:23 some/dir
cgwippern@cgwippern:~/kaniko-dev/issues/704$ tar tvf image-uncached/50e6e4b592a702ddd901bbda27fb0d4f623b91b75312faa84eb6b9e51274a8de/layer.tar --absolute-names
---------- 0/0               0 1969-12-31 16:00 some/dir/.wh.first.txt
drwxr-xr-x /root/root        0 2019-12-22 12:54 /
drwxr-xr-x /root/root        0 2019-12-22 12:54 some
drwxr-xr-x /root/root        0 2019-12-22 12:54 some/dir
```
Specifically the cached version is missing `some/dir/.wh.first.txt` which I think would be used to filter out `some/dir/first.txt` from the final image content. IE because this isn't included in the layer for the cached version of the image `some/dir/first.txt` is included in the final image.

**Comment by cvgw** [comment]
Updating https://github.com/GoogleContainerTools/kaniko/blob/732b2700354dc0f501e5d3012e74b3d6672bae41/pkg/commands/run.go#L209 to `return nil` (as in the case of `RunCommand`) produces the expected behavior but I'm not sure what the consequences of that change would be.

**Comment by cvgw** [comment]
This line is leading to the whiteout file missing in the cached layer
https://github.com/GoogleContainerTools/kaniko/blob/732b2700354dc0f501e5d3012e74b3d6672bae41/pkg/util/fs_util.go#L123

If I understand this correctly, we wouldn't want to extract the whiteout files from base images (which `GetFSFromImage` is also used for) but doing so in the case of a cached layer would make sense.

So perhaps we could have two different versions of `GetFSFromImage` (`GetFSFromLayer`?) or we could pass an option which would wrap line 123 and allow the whiteout files to be extracted.

If the whiteout files are extracted and returned in `extractedFiles` then the rest of the code could remain unchanged.

**Comment by tejal29** [comment]
So, IIUC the cached layer for 
`RUN mkdir -p /some/dir/ && echo 'first' > /some/dir/first.txt` has `/some/dir/first.txt`

Then cached layer for
`RUN rm /some/dir/first.txt` has `/some/dir/.wh.first.txt`

I feel 
1.  `CacheRunCommand` and `RunCommand` should use the same `TakeSnapshotFS` method.
2. I also feel, `CacheRunCommand. ExecuteCommand` should call `util.GetFSFromLayer` instead of https://github.com/GoogleContainerTools/kaniko/blob/732b2700354dc0f501e5d3012e74b3d6672bae41/pkg/commands/run.go#L200
A command always corresponds to a Layer right?

**Comment by cvgw** [comment]
> So, IIUC the cached layer for
RUN mkdir -p /some/dir/ && echo 'first' > /some/dir/first.txt has /some/dir/first.txt
Then cached layer for
RUN rm /some/dir/first.txt has /some/dir/.wh.first.txt

Yup, that's correct

>CacheRunCommand and RunCommand should use the same TakeSnapshotFS method.
I don't see anything wrong with making that change, but I'm not sure if there is a historical reason they aren't doing that currently

>I also feel, CacheRunCommand. ExecuteCommand should call util.GetFSFromLayer instead of
Where is `util.GetFSFromLayer` defined? I wasn't able to find it in the code.

**Comment by tejal29** [comment]
>>> I don't see anything wrong with making that change, but I'm not sure if there is a historical reason they aren't doing that currently

 i feel its safe for us to go ahead and make this change (without worrying about past. Kaniko started as POC and hence there could be simply no reason).
If the tests pass, we should be fine.

>>> Where is util.GetFSFromLayer defined? I wasn't able to find it in the code.

I assumed in your comment [here](https://github.com/GoogleContainerTools/kaniko/issues/704#issuecomment-568327969), `util.GetFSFromLayer` already :) My bad. 
I think there would be very few changes between both like
1. whiteout files 
2. Whitelist paths - maybe?

WDYT?

**Comment by cvgw** [comment]
>I assumed in your comment here, util.GetFSFromLayer is already present :) My bad.
Oh gotcha, that makes sense 😄 

Ya, I think adding another method `util.GetFSFromLayer` that uses the same logic as `util.GetFSFromImage` (do some refactoring to remove duplication) but perhaps takes some options one of which would allow whiteout files to be removed?
